autoPatchelfHook: fix detection under crossSystem

impl opened the pull request:
###### Motivation for this change

In #84415, autoPatchelfHook was taught to use the correct path to the `readelf` binary when a `crossSystem` is specified. Unfortunately, the remainder of the functionality in the script depended on `ldd`, which only reads ELF files of its own architecture. It has the further unfortunate quality of not reporting any useful error, but rather that the file is not a dynamic executable.

This change uses `patchelf` to directly analyze the `DT_NEEDED` tags in the target files instead, which correctly works across architectures. It also updates the use of `objdump` to be prefix-aware `$OBJDUMP` (which would have been required in the PR mentioned above, but we never made it that far into the script execution).

Here is an example of a package that can now be properly cross-patched (?, not sure what the terminology for this is):

```nix
let
  pkgs = import <nixpkgs> {
    crossSystem = {
      config = "armv6l-unknown-linux-gnueabihf";
      gcc = {
        arch = "armv6kz";
        fpu = "vfp";
      };
    };
  };
in pkgs.callPackage ({ lib, stdenv, fetchFromGitHub, autoPatchelfHook, glibc, gccForLibs }: stdenv.mkDerivation rec {
  pname = "vcdbg";
  version = "1.20210831";

  src = fetchFromGitHub {
    owner = "raspberrypi";
    repo = "firmware";
    rev = version;
    sha256 = "sha256-NhQwg6DRGSwsULxXWJAOUPIW+JvOnZVDTTvIBZnjzgA=";
  };

  nativeBuildInputs = [ autoPatchelfHook ];
  buildInputs = [ gccForLibs.lib ];

  installPhase = ''
    install -m 644 -D -t $out/lib opt/vc/lib/*.so
    install -m 755 -D opt/vc/bin/vcdbg $out/bin/vcdbg
  '';

  meta = with lib; {
    description = "vcdbg";
    homepage = "https://github.com/raspberrypi/firmware";
    license = licenses.unfreeRedistributableFirmware; # See https://github.com/raspberrypi/firmware/blob/master/boot/LICENCE.broadcom
  };
}) {}
```

This binary runs correctly even under aarch64 on the Raspberry Pi when built.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

impl commented on pkgs/build-support/setup-hooks/auto-patchelf.sh:
```diff
@@ -54,14 +54,15 @@ declare -gi depCacheInitialised=0
     local so found foundso
     for so in "${autoPatchelfCachedDeps[@]}"; do
-        for found in $(getDepsFromSo "$so"); do
+        for found in $(getDepsFromElfBinary "$so"); do
```
I don't think this change is doing the right thing because the library path isn't evaluated yet. I'll take another pass at this.

impl replied:
OK, I think I've addressed both of these now. Let me know what you think!

SuperSandro2000 commented on pkgs/build-support/setup-hooks/auto-patchelf.sh:
```diff
@@ -54,14 +54,15 @@ declare -gi depCacheInitialised=0
-    ldd "$1" 2> /dev/null | sed -n -e 's/[^=]*=> *\(.\+\) \+([^)]*)$/\1/p'
+getDepsFromElfBinary() {
+    $READELF -d "$1" 2> /dev/null | \
+        sed -n -e 's/^[\t ]*0x[0-9a-f]\{1,\}[\t ]\{1,\}(NEEDED)[\t ]\{1,\}Shared library: \[\([^]]\{1,\}\)\]/\1/p'
```
This regex needs a comment.

impl replied:
I switched this function to just use `patchelf --print-needed` instead which gives exactly the same output.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 137351` at fd5670af run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>50 packages marked as broken and skipped:</summary>
  <ul>
    <li>Literate</li>
    <li>adoptopenjdk-hotspot-bin-13</li>
    <li>adoptopenjdk-hotspot-bin-14</li>
    <li>adoptopenjdk-jre-hotspot-bin-13</li>
    <li>adoptopenjdk-jre-hotspot-bin-14</li>
    <li>cassandra_2_1</li>
    <li>cassandra_2_2</li>
    <li>collectd</li>
    <li>collectd-data</li>
    <li>dtools</li>
    <li>echidna</li>
    <li>gtkd</li>
    <li>impressive</li>
    <li>isabelle</li>
    <li>jboss</li>
    <li>odpdown</li>
    <li>onedrive</li>
    <li>openems</li>
    <li>opentsdb</li>
    <li>ovftool</li>
    <li>pig</li>
    <li>polymake</li>
    <li>processing</li>
    <li>python38Packages.cnvkit</li>
    <li>python38Packages.foundationdb51</li>
    <li>python38Packages.foundationdb52</li>
    <li>python38Packages.foundationdb60</li>
    <li>python38Packages.foundationdb61</li>
    <li>python38Packages.python-csxcad</li>
    <li>python38Packages.python-openems</li>
    <li>python38Packages.tensorflowWithCuda</li>
    <li>python39Packages.cnvkit</li>
    <li>python39Packages.foundationdb51</li>
    <li>python39Packages.foundationdb52</li>
    <li>python39Packages.foundationdb60</li>
    <li>python39Packages.foundationdb61</li>
    <li>python39Packages.jax</li>
    <li>python39Packages.python-csxcad</li>
    <li>python39Packages.python-openems</li>
    <li>python39Packages.tensorflowWithCuda</li>
    <li>qemu_xen</li>
    <li>qemu_xen-light</li>
    <li>qemu_xen_4_10</li>
    <li>qemu_xen_4_10-light</li>
    <li>rund</li>
    <li>spark</li>
    <li>spring</li>
    <li>springLobby</li>
    <li>tilix</li>
    <li>tlaps</li>
  </ul>
</details>
<details>
  <summary>51 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/cd86862503009ecb5d38f071e1c5fe10">adoptopenjdk-bin</a> (adoptopenjdk-hotspot-bin-11 ,openjdk11-bootstrap)</li>
    <li><a href="https://gist.github.com/f234fba6a9e47306c8a08ed507decae5">openjdk8-bootstrap</a> (adoptopenjdk-hotspot-bin-8)</li>
    <li><a href="https://gist.github.com/090ca0a8c76c3d0c33285dd03f83f40b">apacheKafka</a> (apacheKafka_2_8)</li>
    <li><a href="https://gist.github.com/8143a1bfe992ca04336aff70d42173f7">apacheKafka_2_7</a></li>
    <li><a href="https://gist.github.com/a6cce20091bfc2aa60dff978a7b411ce">bazel</a> (bazel_3)</li>
    <li><a href="https://gist.github.com/d8abeb6015537cdb39a9322adaedabdb">bazel_0</a> (bazel_0_26)</li>
    <li><a href="https://gist.github.com/5f2adca5a8ec310213743c7de48611c0">bluej</a></li>
    <li><a href="https://gist.github.com/aaf63eb99259697c471a2149f094dff0">charles</a> (charles4)</li>
    <li><a href="https://gist.github.com/077fe4d3f219066a2828ceea762f82eb">clojure</a></li>
    <li><a href="https://gist.github.com/03cc7cac42fb16f58cb1720ecf821187">cytoscape</a></li>
    <li><a href="https://gist.github.com/378efb2246079a90346f932ae88a1518">dbeaver</a></li>
    <li><a href="https://gist.github.com/ca8e2224d8a03cc7daee67fbbc0d6e3c">dbus_java</a></li>
    <li><a href="https://gist.github.com/63f28df5614fef6d24718e64249717ad">erlang_javac</a></li>
    <li><a href="https://gist.github.com/75a8d2b24d655d818947383a6ce6ed94">filebot</a></li>
    <li><a href="https://gist.github.com/dd511842baf3f6b602ea231b373cdf59">freeplane</a></li>
    <li><a href="https://gist.github.com/096ba74cccafa4666546b2b33286be55">gradle_5</a></li>
    <li><a href="https://gist.github.com/4df436e8631f6bbbfaa90eb6e1a76106">gradle_6</a></li>
    <li><a href="https://gist.github.com/c6a030d9457f880707a66d92da4cc2a8">graylog</a></li>
    <li><a href="https://gist.github.com/f29d6e2adfb29b8468022d71789db166">greenfoot</a></li>
    <li><a href="https://gist.github.com/3184c6411ffa6be5fc313c31ef0aa13c">gremlin-console</a></li>
    <li><a href="https://gist.github.com/9f8162ef5b479db3c5b321a718a9ee77">igv</a></li>
    <li><a href="https://gist.github.com/5d0221b99e9b00a8a6edaafe438c2a74">jdk11</a> (openjdk11)</li>
    <li><a href="https://gist.github.com/3649acab7def6635bace115c62bfb7c5">jdk11_headless</a> (openjdk11_headless)</li>
    <li><a href="https://gist.github.com/9b943dac6717dab2f457f99c5abb3d7f">jetbrains.clion</a></li>
    <li><a href="https://gist.github.com/9015436c01380c0084abb7dc26393bae">jetbrains.datagrip</a></li>
    <li><a href="https://gist.github.com/b7d8b6b7209ec36ef2bc1d84c03e0648">jetbrains.goland</a></li>
    <li><a href="https://gist.github.com/8c47279dbf3a25a9117c955862b29bdb">jetbrains.jdk</a></li>
    <li><a href="https://gist.github.com/ef2ce1da39429563ce085cf526cba1ff">jetbrains.mps</a></li>
    <li><a href="https://gist.github.com/636bcefc252b5cc842b719a61e06bb61">jetbrains.phpstorm</a></li>
    <li><a href="https://gist.github.com/6dce312614291158bab91666c1671571">jetbrains.pycharm-community</a></li>
    <li><a href="https://gist.github.com/e0b0247d0ab65d2bde92b6b9f40bb4ef">jetbrains.pycharm-professional</a></li>
    <li><a href="https://gist.github.com/330d16412964e128ebe178478f97dbf1">jetbrains.rider</a></li>
    <li><a href="https://gist.github.com/a8af5f2e3ea9a9cfa78c12dbf3dd09e6">jetbrains.ruby-mine</a></li>
    <li><a href="https://gist.github.com/5039039cca916295499316a4c1dae1f5">jetbrains.webstorm</a></li>
    <li><a href="https://gist.github.com/de9578983ec4de9712162c6f36ee1420">jre8</a> (jre8_headless)</li>
    <li><a href="https://gist.github.com/edc23c8b9fe660f6e9d9f4590339266a">libmatthew_java</a></li>
    <li><a href="https://gist.github.com/bea2d36345a6953873f518844a0976a0">logstash</a> (logstash6)</li>
    <li><a href="https://gist.github.com/eb67aac461cda96d5d2ad31bb947caba">logstash6-oss</a></li>
    <li><a href="https://gist.github.com/9818ecccd024074af17e8ceab2bacd20">metabase</a></li>
    <li><a href="https://gist.github.com/2e71a99441dc817786339e296046d323">mucommander</a></li>
    <li><a href="https://gist.github.com/86c610827ce95cc325039b7265dd7604">netbeans</a></li>
    <li><a href="https://gist.github.com/f4d432a598db034dc23c78fd24266981">nzbhydra2</a></li>
    <li><a href="https://gist.github.com/388c27a7a3af0e92aac7565e51e56300">pdfsam-basic</a></li>
    <li><a href="https://gist.github.com/f612d5804c4fb9af1eb23e18f8e73558">python38Packages.adb-enhanced</a></li>
    <li><a href="https://gist.github.com/39661bf7bb524d8b6dd64422a197beab">python39Packages.adb-enhanced</a></li>
    <li><a href="https://gist.github.com/06234b2c735814277eedd851f6777cde">scenebuilder</a></li>
    <li><a href="https://gist.github.com/c15435b4c08f52db78ce87b5f6acfcf2">signal-cli</a></li>
    <li><a href="https://gist.github.com/e806959423f4ee81223e8713b4ecb982">smartgithg</a></li>
    <li><a href="https://gist.github.com/1bfbad0414f70012f3870b98f545c473">soapui</a></li>
    <li><a href="https://gist.github.com/8dbf139c7cf8c7675ed0cca250d7adae">stm32cubemx</a></li>
    <li><a href="https://gist.github.com/1b314523895d780fdfb27a990b0d28c4">tlaplus</a></li>
  </ul>
</details>
<details>
  <summary>479 packages skipped due to time constraints:</summary>
  <ul>
    <li>DisnixWebService</li>
    <li>R</li>
    <li>_1password</li>
    <li>_7zz</li>
    <li>abcl</li>
    <li>adapta-gtk-theme</li>
    <li>openjdk16-bootstrap (adoptopenjdk-hotspot-bin-15)</li>
    <li>adoptopenjdk-hotspot-bin-16</li>
    <li>adoptopenjdk-icedtea-web</li>
    <li>adoptopenjdk-jre-bin (adoptopenjdk-jre-hotspot-bin-11)</li>
    <li>...</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

impl committed:
autoPatchelfHook: fix detection under crossSystem

In #84415, autoPatchelfHook was taught to use the correct path to the
readelf binary when a crossSystem is specified. Unfortunately, the
remainder of the functionality in the script depended on ldd, which only
reads ELF files of its own architecture. It has the further unfortunate
quality of not reporting any useful error, but rather that the file is
not a dynamic executable.

This change uses patchelf to directly analyze the DT_NEEDED tags in the
target files instead, which correctly works across architectures. It
also updates the use of objdump to be prefix-aware $OBJDUMP (which would
have been required in the PR mentioned above, but we never made it that
far into the script execution).

symphorien reviewed (approved):
could build julia_16-bin natively and cross to aarch64. native julia passes its tests, cross julia can run --version in qemu.

impl commented:
Just to let you know, I have a few more fixes I'll put into a follow-up PR (I rented a big ol' AWS instance and manually reviewed all of the output of `nixpkgs-review pr`... :) that should improve stability across existing packages as well. Coming shortly!

impl commented:
That PR is #137886, for future onlookers!
